### DISCUSSION TITLE
cbc standalone binary file

### DISCUSSION THREAD

**Discussion by LeeU1911** [coin-or/Cbc#409] [Discussion]
Hey there,

I tried to download packages from https://www.coin-or.org/download/binary/Cbc/?C=M;O=D but it doesn't seem to have a single cbc binary file. Instead the `bin` folder has `cbc`, `clp` and `glpsol` which are all have sizes in KB. If I run `cbc <some_lp_file>` from this folder, it works, but it wouldn't work if I use the `cbc` file alone.

We have an older version 2.9.9 of cbc standalone that is 7MB in size (didn't put in place by me) so I want to upgrade to newer version. Can you recommend a way to download the standalone `cbc` binary file for linux?

Thanks a lot!

**Comment by tkralphs** [comment]
Cbc can be built straightforwardly on Linux, so the best thing would be to just build your own. From the README, this will install the latest release:
```
wget https://raw.githubusercontent.com/coin-or/coinbrew/master/coinbrew
chmod u+x coinbrew
./coinbrew fetch Cbc@2.10.5
./coinbrew build Cbc
```
The second-best thing would be to install Cbc using a package manager. Instructions for that can also be found in the README.

If you really want to download binaries, that's the right place to look (for now), but you'll have to provide more information in order for us to help. 

- What OS and version are you on? 
- What exact file did you download? 
- When you say cbc doesn't work if you use the `cbc` file alone, what was the error? 

The reason for the difference in size of the executables is most likely that the executable you currently have is statically linked, while what you downloaded is probably linking to shared libraries and is therefore much smaller.

**Reply by LeeU1911** [reply]
Hey Ted,

Thanks a lot for your detailed reply. Yes I think the one with large size is statically linked executable. Let me try to build using `coinbrew` to see if I can achieve that.

A bit more details of targeted environment:
- CBC current version 2.9.9. I'm upgrading to 2.10.5 for faster performance
- Project runs in serverless deployment (AWS Lambda deployed with Zappa) running Python 3.6 in Amazon Linux OS

I downloaded this file earlier: https://www.coin-or.org/download/binary/Cbc/Cbc-master-precise-linux-x86_64-gcc9.tgz but it's a shared libraries linked version as you pointed out.

**Reply by LeeU1911** [reply]
Hey @tkralphs, I have tried with your commands but couldn't build a static executable even with `--static` with `coinbrew` as `./coinbrew build Cbc --static`

I found this old [thread](https://github.com/coin-or/Cbc/issues/256) of yours and followed the commands mentioned there but also no success. After trying `libtool` command, `cbc` output file is only 11KB in size and there's a warning `libtool: warning: complete static linking is impossible in this configuration`. Any idea? Thanks.

**Reply by tkralphs** [reply]
Sorry, I lost track of the fact that this was unresolved. Do you need a completely statically linked executable? Or just a package that can be relocated and will work on any Linux box? Building a completely statically linked executable is relatively easy with the current `master` version of Cbc, but is harder with `2.10.*` and may involve some hand configuration after building and looking at the results of `ldd cbc` in that case. If you just need a statically linked executable and nothing else, you may save yourself a lot of trouble by just grabbing one from [AMPL](https://ampl.com/products/solvers/open-source/#cbc) (the version is not quite the latest, but should be fine). If you really need to build one yourself, I can try to help, but the instructions in the thread you pointed to are the best starting point I can recommend. It will be a lot easier if you focus on `master`, which will be Cbc 3.0 "soon."

**Reply by LeeU1911** [reply]
Hey @tkralphs, thanks for your reply.  I need a statically linked executable so I can put into my program running as serverless function on AWS Lamdba. I couldn’t use AMPL because I need version higher than 2.10.3 due to this [issue](https://github.com/coin-or/Cbc/issues/286). It seems to resolve in 2.10.5 and master as far as I tested with dynamically linked executable. 

Everytime I try `./coinbrew build Cbc —static` followed with `libtool` command, it gives warning as mentioned. Is there anything you can suggest for me to try differently?

Also, do you know why we stopped building statically linked executable for each release? Was hoping to find one for 2.10.5 but couldn’t find anywhere (AMPL, bintray, linux package manager, etc.)

**Reply by tkralphs** [reply]
My apologies for the confusion, it seems that the fix to our build tools mentioned in the thread referenced above, that was needed to make the static build work with `stable/2.10`, somehow did not make it into the `2.10.5` release, so I haven't been able to successfully make a completely static executable with that release. I _was_ successful with the current `stable/2.10` and dependencies. I will try to make a `2.10.6` with this fix, but in the meantime, I was able to get a static executable with more or less the same recipe from the thread. Here is what I did just now that worked.
```
~> coinbrew fetch Cbc@2.10
~> coinbrew  build Cbc --no-third-party --enable-static --disable-shared --disable-bzlib --disable-readline --disable-zlib --without-cholmod --without-lapack --without-amd LT_LDFLAGS=-all-static CXXFLAGS="-static -static-libgcc -static-libstdc++" 
~> cd dist/bin
~> ldd cbc
        not a dynamic executable
```
This recipe is a bit fragile and may not work universally, which is part of the reason we stopped trying to build statically with `stable/2.10` (I don't know about AMPL, there is no obvious reason they haven't made a static build of 2.10.5). 

Static builds with current master are much more robust and the recipe is simpler due to the use of a more recent version of the autotools. Hope this helps for now, I will update if/when there is a `2.10.6`.

**Reply by LeeU1911** [reply]
Thanks @tkralphs. I finally was able to build a statically linked executable with `stable/2.10` followed your commands.
Though what we wanted to upgrade was 2.10.3 where there was a good performance improvement for our use case.

I also tried on `master` and successful with the same recipe. Thanks a lot for your help. 

Just an extra question, do you know if the performance improvements in `2.10.3` made it to current master? It seems to me that `master` took `10`+ seconds to solve the lp mentioned [here](https://github.com/coin-or/Cbc/issues/286#issuecomment-588702974) while `stable/2.10` took only `7`+ seconds. But I don't have much clue about how solvers work, so wanted to double check

Thanks again!
Vu

**Reply by tkralphs** [reply]
The performance of `master` should be better overall than that of 2.10.3, but on any one instance, anything can happen. I think the instance you're referring to is a pure LP and as far as I know, the version of Clp (the underlying LP solver) used by 2.10.3 and the version used by `master` are not very different (if at all). There is probably more of an emphasis in tuning to focus on instances that are very difficult (taking minutes or hours) and not worry as much about instances that take seconds. At times, there may be some modifications to the solver that are an improvement overall, but that lead to regressions for easier instances.

**Reply by LeeU1911** [reply]
Thanks for the additional info Ted. Really appreciate it!

Best, Vu
